pkg/client/apiutil/apimachinery.go

```diff
@@ -135,6 +135,13 @@ func GVKForObject(obj runtime.Object, scheme *runtime.Scheme) (schema.GroupVersi
 	// Use the given scheme to retrieve all the GVKs for the object.
 	gvks, isUnversioned, err := scheme.ObjectKinds(obj)
 	if err != nil {
+		if runtime.IsNotRegisteredError(err) && obj.GetObjectKind() != nil {
+			// Use object's current GVK as a last ditch attempt for a type not registered with the scheme.
+			currentGVK := obj.GetObjectKind().GroupVersionKind()
+			if !currentGVK.Empty() {
+				return currentGVK, nil
+			}
+		}
```

Comment:
🤔 Reading through the use case, it makes sense why one would want to reuse a type across many types and ultimately make use of the facilities already in place.

This code IIRC is used in a number of places, including the client itself, by allowing non-registered types to pass through, aren't now allowing types to go through caches/clients even when they're not supported?

Reply:
> This code IIRC is used in a number of places, including the client itself, by allowing non-registered types to pass through, aren't now allowing types to go through caches/clients even when they're not supported?

My desire is to make those types behave largely as if they are supported. To do so, I currently wrap the manager returned client adding a check if the obj is registered with the scheme, and if not converts it to an unstructured before calling the upstream client, and then reflects any updated values on the unstructured object back. (This behavior is largely working in the latest release of [reconciler-runtime](https://github.com/vmware-labs/reconciler-runtime))

Updating `GVKForObject` will help avoid conversions to/from unstructured in other parts of the codebase that are harder to wrap and user code, like `SetControllerReference`.

> Example, when creating a controller `GVKForObject` is used to determine the GVK, and we pass in the scheme given that we're expecting to use the scheme to create new instances of that object, if that object isn't registered, it won't fail until much later?

I've never seen someone use a non-empty object as the type, but it's certainly possible. If you like, I can add a defensive check in the builder that will fail if the object is not unstructured or registered in the scheme.

Reply:
I don't think as of today we ever assume that a type that is not unstructured/PartialObjectMeta can pass this check and not be registered in the scheme. So this definitely has potential to cause issues both for us and existing users of the downstream code.

What is the reason to not simply do this downstream through a slim wrapper around this?

Reply:
> I don't think as of today we ever assume that a type that is not unstructured/PartialObjectMeta can pass this check and not be registered in the scheme.

Correct, it always errs today.

>What is the reason to not simply do this downstream through a slim wrapper around this?

It's not easy to wrap bare functions, unlike an interface that is expected to be injected (e.g. `Client`). I'd have to provide alternate implementation and then tell users to call my custom api, or resort to nasty hacks.

Reply:
Not sure I follow. So you can tell your users to use a different type but not to use a different way of inferring the GVK from that type?

> It's not easy to wrap bare functions, 
It is:
```
func wrapping(f foo, b bar)(gvk, err)
  upstreamResult, err := upstream(f, b)
  if err != nil {
     return doSomethingElse(f, b)
  }
  return upstreamResult, nil
}
```

Reply:
@alvaroaleman yea, that's the easy part. The hard part is switching every caller to use the wrapper function.

Reply:
A random thought about this: maybe we could do something like Go has done for errors, with Wrap/Unwrap, and have a fallback to check a well-defined interface to Unwrap the true type of a struct. This would give us the ability to keep the current behavior, while introducing this new functionality of common types.